### PR TITLE
Provide mechanism for sharing file regions cache with SarifLogger instance.

### PR DESCRIPTION
**Pull request by michaelcfanning**
There is currently no mechanism for an upstream caller to initialize a `SarifLogger` instance with a specific file regions cache.

This limits `SarifLogger` utilization for more unusual API scenarios.

@eddynaka

### CODE REVIEW

File: src/Sarif/FileRegionsCache.cs [hunk 1]
```diff
@@ -142,15 +142,15 @@ private Region PopulateTextRegionProperties(NewLineIndex lineIndex, Region input
             return region;
         }
 
-        public Region ConstructMultilineContextSnippet(Region inputRegion, Uri uri)
+        public Region ConstructMultilineContextSnippet(Region inputRegion, Uri uri, string fileText = null)
```
Comment (michaelcfanning) [inline src/Sarif/FileRegionsCache.cs:145]: >ConstructMultilineContextSnippet [](http://example.com/codeflow?start=22&length=32)

Non-breaking change.

File: src/Sarif/Visitors/InsertOptionalDataVisitor.cs [hunk 2]
```diff
@@ -32,8 +32,14 @@ public class InsertOptionalDataVisitor : SarifRewritingVisitor
         private const string Email = nameof(Email);
         private const string CommitSha = nameof(CommitSha);
 
-        public InsertOptionalDataVisitor(OptionallyEmittedData dataToInsert, Run run, IEnumerable<string> insertProperties)
-            : this(dataToInsert, run?.OriginalUriBaseIds, insertProperties: insertProperties)
+        public InsertOptionalDataVisitor(OptionallyEmittedData dataToInsert,
```
Comment (michaelcfanning) [inline src/Sarif/Visitors/InsertOptionalDataVisitor.cs:35]: >InsertOptionalDataVisitor [](http://example.com/codeflow?start=15&length=25)

This helper did not previously allow for specifying the file regions cache. Our overloads in this type are unfortunate.

The problem is specifying original uri base ids independently vs. providing a run instance...

File: src/Sarif/Writers/SarifLogger.cs [hunk 3]
```diff
@@ -112,32 +138,7 @@ public class SarifLogger : BaseLogger, IDisposable, IAnalysisLogger
                 {
                     RuleToIndexMap[_run.Tool.Driver.Rules[i]] = i;
                 }
-            }
-        }
-
-        private SarifLogger(TextWriter textWriter,
```
Comment (michaelcfanning) [inline src/Sarif/Writers/SarifLogger.cs:118]: >SarifLogger [](http://example.com/codeflow?start=16&length=11)

I zapped this ctor.

File: src/Sarif/Writers/BaseLogger.cs [hunk 4]
```diff
@@ -15,9 +15,15 @@ public abstract class BaseLogger
         protected BaseLogger(IEnumerable<FailureLevel> failureLevels,
             IEnumerable<ResultKind> resultKinds)
         {
-            _failureLevels = failureLevels?.ToList() ?? new List<FailureLevel>();
+            _failureLevels = failureLevels?.ToList() ?? new List<FailureLevel>(); ;
             _resultKinds = resultKinds?.ToList() ?? new List<ResultKind>();
 
+            if (_failureLevels.Count == 0 && _resultKinds.Count == 0)
```
Comment (michaelcfanning) [inline src/Sarif/Writers/BaseLogger.cs:21]: >_failureLevels [](http://example.com/codeflow?start=16&length=14)

This is just a helper, if you provide us entirely empty data, we will provide you sensible defaults.

File: src/Sarif/Writers/BaseLogger.cs [hunk 5]
```diff
@@ -15,9 +15,15 @@ public abstract class BaseLogger
         protected BaseLogger(IEnumerable<FailureLevel> failureLevels,
             IEnumerable<ResultKind> resultKinds)
         {
-            _failureLevels = failureLevels?.ToList() ?? new List<FailureLevel>();
+            _failureLevels = failureLevels?.ToList() ?? new List<FailureLevel>(); ;
```
Comment (cfaucon) [inline src/Sarif/Writers/BaseLogger.cs:18]: ?

**Pull request merged**